File: android-app/app/src/main/java/com/arsvechkarev/vault/viewbuilding/TypefaceSpan.java
package com.arsvechkarev.vault.viewbuilding;

import android.graphics.Paint;
import android.graphics.Typeface;
import android.text.TextPaint;
import android.text.style.MetricAffectingSpan;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Span that updates the typeface of the text it's attached to. The
 * <code>TypefaceSpan</code> can be constructed either based on a font family or based on
 * a <code>Typeface</code>. When {@link #TypefaceSpan(String)} is used, the previous style
 * of the <code>TextView</code> is kept. When {@link #TypefaceSpan(Typeface)} is used,
 * the
 * <code>Typeface</code> style replaces the
 * <code>TextView</code>'s style.
 * <p>
 * For example, let's consider a <code>TextView</code> with
 * <code>android:textStyle="italic"</code> and a typeface created based on a font from
 * resources, with a bold style. When applying a <code>TypefaceSpan</code> based the
 * typeface, the text will only keep the bold style, overriding the
 * <code>TextView</code>'s textStyle. When applying a
 * <code>TypefaceSpan</code> based on a font family: "monospace", the resulted text will
 * keep the italic style.
 * <pre>
 * Typeface myTypeface = Typeface.create(ResourcesCompat.getFont(context, R.font.acme),
 * Typeface.BOLD);
 * SpannableString string = new SpannableString("Text with typeface span.");
 * string.setSpan(new TypefaceSpan(myTypeface), 10, 18, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
 * string.setSpan(new TypefaceSpan("monospace"), 19, 22, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
 * </pre>
 * <img src="{@docRoot}reference/android/images/text/style/typefacespan.png" />
 * <figcaption>Text with <code>TypefaceSpan</code>s constructed based on a font from
 * resource and from a font family.</figcaption>
 */
public class TypefaceSpan extends MetricAffectingSpan {

    private final String mFamily;

    private final Typeface mTypeface;

    /**
     * Constructs a {@link TypefaceSpan} based on the font family. The previous style of the
     * TextPaint is kept. If the font family is null, the text paint is not modified.
     *
     * @param family The font family for this typeface.  Examples include "monospace",
     *               "serif", and "sans-serif"
     */
    public TypefaceSpan(@Nullable String family) {
        this(family, null);
    }

    /**
     * Constructs a {@link TypefaceSpan} from a {@link Typeface}. The previous style of the
     * TextPaint is overridden and the style of the typeface is used.
     *
     * @param typeface the typeface
     */
    public TypefaceSpan(@NonNull Typeface typeface) {
        this(null, typeface);
    }

    private TypefaceSpan(@Nullable String family, @Nullable Typeface typeface) {
        mFamily = family;
        mTypeface = typeface;
    }

    @Override
    public void updateDrawState(@NonNull TextPaint ds) {
        updateTypeface(ds);
    }

    @Override
    public void updateMeasureState(@NonNull TextPaint paint) {
        updateTypeface(paint);
    }

    private void updateTypeface(@NonNull Paint paint) {
        if (mTypeface != null) {
            paint.setTypeface(mTypeface);
        } else if (mFamily != null) {
            applyFontFamily(paint, mFamily);
        }
    }

    private void applyFontFamily(@NonNull Paint paint, @NonNull String family) {
        int style;
        Typeface old = paint.getTypeface();
        if (old == null) {
            style = Typeface.NORMAL;
        } else {
            style = old.getStyle();
        }
        final Typeface styledTypeface = Typeface.create(family, style);
        int fake = style & ~styledTypeface.getStyle();

        if ((fake & Typeface.BOLD) != 0) {
            paint.setFakeBoldText(true);
        }

        if ((fake & Typeface.ITALIC) != 0) {
            paint.setTextSkewX(-0.25f);
        }
        paint.setTypeface(styledTypeface);
    }
}

File: android-app/app/src/main/java/com/arsvechkarev/vault/core/TypefaceSpan.kt
package com.arsvechkarev.vault.core

import android.graphics.Paint
import android.graphics.Typeface
import android.text.TextPaint
import android.text.style.MetricAffectingSpan

/*
 * ========================================================
 * =============== COPIED FROM ANDROID SDK  ===============
 * ========================================================
 */
/**
 * Span that updates the typeface of the text it's attached to. The
 * `TypefaceSpan` can be constructed either based on a font family or based on
 * a `Typeface`. When [TypefaceSpan] is used, the previous style
 * of the `TextView` is kept. When [TypefaceSpan] is used,
 * the
 * `Typeface` style replaces the
 * `TextView`'s style.
 *
 *
 * For example, let's consider a `TextView` with
 * `android:textStyle="italic"` and a typeface created based on a font from
 * resources, with a bold style. When applying a `TypefaceSpan` based the
 * typeface, the text will only keep the bold style, overriding the
 * `TextView`'s textStyle. When applying a
 * `TypefaceSpan` based on a font family: "monospace", the resulted text will
 * keep the italic style.
 * <pre>
 * Typeface myTypeface = Typeface.create(ResourcesCompat.getFont(context, R.font.acme),
 * Typeface.BOLD);
 * SpannableString string = new SpannableString("Text with typeface span.");
 * string.setSpan(new TypefaceSpan(myTypeface), 10, 18, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
 * string.setSpan(new TypefaceSpan("monospace"), 19, 22, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
</pre> *
 * <img src="{@docRoot}reference/android/images/text/style/typefacespan.png"></img>
 * <figcaption>Text with `TypefaceSpan`s constructed based on a font from
 * resource and from a font family.</figcaption>
 */
/*
 * ========================================================
 * =============== COPIED FROM ANDROID SDK  ===============
 * ========================================================
 */
class TypefaceSpan(
  private val mTypeface: Typeface?,
  private val mFamily: String? = null,
) : MetricAffectingSpan() {
  
  override fun updateDrawState(ds: TextPaint) {
    updateTypeface(ds)
  }
  
  override fun updateMeasureState(paint: TextPaint) {
    updateTypeface(paint)
  }
  
  private fun updateTypeface(paint: Paint) {
    if (mTypeface != null) {
      paint.setTypeface(mTypeface)
    } else {
      mFamily?.let { family -> applyFontFamily(paint, family) }
    }
  }
  
  private fun applyFontFamily(paint: Paint, family: String) {
    val old = paint.typeface
    val style = old?.style ?: Typeface.NORMAL
    val styledTypeface = Typeface.create(family, style)
    val fake = style and styledTypeface.style.inv()
    if (fake and Typeface.BOLD != 0) {
      paint.isFakeBoldText = true
    }
    if (fake and Typeface.ITALIC != 0) {
      paint.textSkewX = -0.25f
    }
    paint.setTypeface(styledTypeface)
  }
}
